_upgrading/sopel-8.md

```diff
@@ -0,0 +1,435 @@
+
+In Sopel 8.0, the `@sopel.plugin.require_privilege()` decorator now implies
+`@sopel.plugin.require_chanmsg()`, and the decorated callable will not run if
+triggered outside of a channel. Previously, `require_privilege` restrictions
+were simply ignored, and the callable would run anyway.
+
+The `@sopel.plugin.require_bot_privilege()` decorator also now implies
+`@sopel.plugin.require_chanmsg()`, with the same associated behavior change.
+
+### Capability negotiation rework
+
+The old `CapReq` method of asking Sopel to negotiate additional capabilities on
+behalf of your plugin has been replaced with a much more robust system based on
+the new `sopel.plugin.capability` class/decorator.
+
+Plugins can now request capabilities in one of two ways:
+
+```python
+"""Sample plugin file"""
+
+from sopel import plugin
+
+# this will register a capability request
+CAP_ACCOUNT_TAG = plugin.capability('account-tag')
+
+# this will work as well
+@plugin.capability('message-prefix')
+def cap_message_prefix(cap_req, bot, acknowledged):
+    # do something if message-prefix is ACK or NAK
+    ...
+```
+
+The first method is easier if you just want to require a capability. The second
+is easier if your plugin needs to _do something in response to_ the IRC server's
+response to the capability request. See the documentation for more details.
+
+### The bot's hostmask
+
+In Sopel 8, accessing `bot.hostmask` when the bot lacks sufficient data to
+determine its hostmask will return `None`. This replaces the previous behavior,
+which was to raise a `KeyError`.
+
+### Testing tool changes
+
+* Convenience methods of the mock IRC server available through Sopel's `pytest`
+  plugin implement keyword-only arguments now. This applies to:
+  * `MockIRCServer.channel_joined()`
+  * `MockIRCServer.mode_set()`
+  * `MockIRCServer.join()`
+  * `MockIRCServer.say()`
+  * `MockIRCServer.pm()`
+
+### Moved API features
+
+We wanted to organize things better in Sopel 8.0, so some of the API features
+have moved. The old locations will still work until Sopel 9.0, but you'll be way
+ahead of the game if you update your plugins now!
+
+* `Identifier` type moved from `sopel.tools` to `sopel.tools.identifiers`
+* `SopelMemory` and `SopelMemoryWithDefault` types moved from `sopel.tools` to
+  `sopel.tools.memories`
+* `sopel.tools.check_pid()` moved to `sopel.cli.utils`
+
+### Removed API features
+
+Previously deprecated parts of Sopel's API have been removed, including:
+
+* `bot.memory['url_callbacks']` is no longer created or populated by the bot
+  * Tracking this information moved to an internal property, in preparation for
+    the removal of `bot.register_url_callback()`, `bot.search_url_callbacks()`,
+    and `bot.unregister_url_callback()` in Sopel 9.0
+  * As a reminder, plugins **should** no longer use the above-mentioned methods;
+    the `bot.rules.check_url_callbacks(bot, url)` method now works very well
+    with the `@sopel.plugin.url()` decorator
```

Comment:
```md
* `bot.memory['url_callbacks']` is no longer created or populated by the bot
  * `bot.register_url_callback()` and `bot.unregister_url_callback()` are deprecated.
  * Use [`@sopel.plugin.url()`][url-decorator] or [`@sopel.plugin.url_lazy()`][url-lazy-decorator] instead.
  * `bot.search_url_callbacks()` is deprecated, replace with [`bot.rules.check_url_callbacks(bot, url)`][check-url-callbacks]
    
  [url-decorator]: https://docs.sopel.chat/plugin/decorators.html#sopel.plugin.url
  [url-lazy-decorator]: https://docs.sopel.chat/plugin/decorators.html#sopel.plugin.url_lazy
  [check-url-callbacks]: https://docs.sopel.chat/package/plugins/rules.html#sopel.plugins.rules.Manager.check_url_callback
```

The last doc link is... not super helpful for understanding you need to do `bot.rules...`

Reply:
I adapted much of this. Couldn't apply directly anyway, for multiple reasons.